pkg/builder/searcherclient/searcher_client.go

```diff
@@ -0,0 +1,226 @@
+}
+
+func NewSearcherClient(logger *slog.Logger, inclusionProof bool) SearcherClient {
+	return &searcherClient{
+		searchers:            make(map[string]Searcher),
+		logger:               logger,
+		inclusionProofActive: inclusionProof,
+	}
+}
+
+func (s *searcherClient) AddSearcher(searcher Searcher) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.searchers[searcher.ID()] = searcher
+	delete(s.disconnected, searcher.ID())
+}
+
+func (s *searcherClient) RemoveSearcher(id string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	searcher, found := s.searchers[id]
+	if !found {
+		return
+	}
+	err := searcher.Close()
+	if err != nil {
+		s.logger.Error("failed to close searcher", "err", err)
+	}
+	delete(s.searchers, searcher.ID())
+}
+
+func (s *searcherClient) Disconnected(id, reason string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.disconnected[id] = reason
+	delete(s.searchers, id)
+}
+
+func (s *searcherClient) SubmitBlock(
+	ctx context.Context,
+	payload *capella.SubmitBlockRequest,
+) error {
+	if payload == nil || len(payload.ExecutionPayload.Transactions) == 0 {
+		return nil
+	}
+
+	minTipTxn, maxTipTxn := big.NewInt(0), big.NewInt(0)
+	ts := time.Unix(int64(payload.ExecutionPayload.Timestamp), 0).Format(time.RFC1123)
+	baseFee := binary.LittleEndian.Uint32(payload.ExecutionPayload.BaseFeePerGas[:])
+
+	blockMetadata := SuperPayload{
+		InternalMetadata: Metadata{
+			RecTimestamp: time.Now(),
+			Builder:      payload.Message.BuilderPubkey.String(),
+			Number:       int64(payload.ExecutionPayload.BlockNumber),
+			BlockHash:    payload.Message.BlockHash.String(),
+			Timestamp:    ts,
+			BaseFee:      baseFee,
+			Transactions: Transaction{
+				Count: int64(len(payload.ExecutionPayload.Transactions)),
+			},
+		},
+	}
+	if s.inclusionProofActive {
+		blockMetadata.SearcherTxns = make(map[string][]string)
+	}
+
+	for idx, btxn := range payload.ExecutionPayload.Transactions {
+		var txn types.Transaction
+		err := txn.UnmarshalBinary(btxn)
+		if err != nil {
+			s.logger.Error("failed to decode transaction", "err", err)
+			continue
+		}
+		// Extract Min/Max
+		if txn.GasTipCap().Cmp(minTipTxn) < 0 || idx == 0 {
+			minTipTxn = txn.GasTipCap()
+		}
+		if txn.GasTipCap().Cmp(maxTipTxn) > 0 || idx == 0 {
+			maxTipTxn = txn.GasTipCap()
+		}
+
+		from, err := types.Sender(types.LatestSignerForChainID(txn.ChainId()), &txn)
+		if err != nil {
+			s.logger.Error("failed to decode sender of transaction",
+				"err", err,
+				"txn", txn.Hash().String(),
+			)
+			continue
+		}
+		clientID := from.Hex()
+		if s.inclusionProofActive {
+			blockMetadata.SearcherTxns[clientID] = append(
+				blockMetadata.SearcherTxns[clientID],
+				txn.Hash().String(),
+			)
+		}
+	}
+
+	blockMetadata.InternalMetadata.Transactions.MinPriorityFee = minTipTxn.Int64()
+	blockMetadata.InternalMetadata.Transactions.MaxPriorityFee = maxTipTxn.Int64()
+
+	s.logger.Info("block metadata processed", "blockMetadata", blockMetadata)
+
+	if s.inclusionProofActive {
```

Comment:
Longer term, I was thinking we decouple the inclusionProof section from the hints payload (the section above this conditional block). 
Basically defer this processing, since it's pretty heavy duty. compared to the above data processing which only needs to happen once, and can be copied to N searchers directly.